### ISSUE TITLE
🔷 [ProjectTracking] Enable mainnet validators to track a single shard

### ISSUE THREAD

**Issue by VanBarbascu** [near/nearcore#10462]
# Goals

## Background

In our current setup, in `mainnet`, every node tracks all shards. State Sync work enabled nodes to easily download state data, unblocking them from tracking a single shard. However, initial testing showed bugs and limitations outside of StateSync code, and this issue tracks the effort to address them and fully enable mainnet nodes in tracking a single shard.

## Why should NEAR One work on this

- Single shard tracking contributes to network scalability and reduces the hardware requirements for chunk only producers.
- Stake Wars IV will run in a network (Statelessnet) where nodes track a subset of shards.

## What needs to be accomplished

We need to fix missing incoming receipts for shards not tracked. State sync implementation has a dependency on nodes tracking all shards. When we generate the state sync header for a shard (call it `T`), a node looks at all the other shards for incoming receipts for shard `T` and adds them to the header. This is not possible if a node only tracks a subset of shards. More info in the [doc](https://docs.google.com/document/d/1ERF6nNrCNeqkMQPeWOqYm7o5eG28HLmDBZVMuSYXe8I/edit). Initially, we will fix GCS-based State Sync by uploading the header from a host that tracks all shards.  The rest of the hosts can download the header from the external location. For decentralised State Sync we will add a new work item to propose a solution. This will enable stateless validation work to go in parallel with this effort.

## Main use case

The main usecase for enabling single shard tracking is Stateless validation.

# Links to external documentations and discussions

[doc](https://docs.google.com/document/d/1ERF6nNrCNeqkMQPeWOqYm7o5eG28HLmDBZVMuSYXe8I/edit)

# Estimated effort

Engineers assigned: @VanBarbascu , @telezhnaya.

Initial effort estimate is about 0.5 PM (person months): transferring headers via GCS
Currently remaining effort is presented in the latest comment of this issue.

# Assumptions

In the single shard tracking scenario, in the next months, RPC nodes will still be able to track all shards of mainnet.

# Pre-requisites

N/A

# Out of scope

Single shard tracking with decentralized state sync. This involves more work and will be addressed in a different issue. 
# Task list:

```[tasklist]
### Tasks
- [ ] https://github.com/near/nearcore/issues/10459
- [x] Read header from external storage
- [x] Update metrics for uploaded/downloaded headers
- [ ] https://github.com/near/nearcore/issues/10570
```

**Comment by VanBarbascu** [comment]
**30 Jan 2024**

Upload and download feature for state sync header is merged and tested in mainnet and testnet. 
Parts are uploaded in GCP i.e. [testnet](https://console.cloud.google.com/storage/browser/state-parts/chain_id%3Dmainnet/epoch_height%3D2353/epoch_id%3DHAuJng8qLkrD97aEAan3Jxm6gbkcyjpHCYLTRhNefEfy;tab=objects?pageState=(%22StorageObjectListTable%22:(%22f%22:%22%255B%255D%22))&project=pagoda-public-services&prefix=&forceOnObjectsSortingFiltering=false). Mainnet can be found by changing paths.
In [this Grafana](https://nearinc.grafana.net/goto/Z7rVO0pSR?orgId=1) dash, in the selected interval I had 2 nodes (mainnet and testnet) syncing from external storage using the new download flow.

**Comment by VanBarbascu** [comment]
Stateless validation with single shard tracking is unblocked with sync from external storage.
We need to tweak some existing config parameters to make sure that we get optimal performance for statelessnet.

### Next: 
- Fix monitoring for uploaded files to check state sync header in the external bucket.

### After we release of DSS feature:
- In the context of peer state sync, we need to have a way of building the state sync header from multiple peers tracking a subset of shards.

**Comment by VanBarbascu** [comment]
This week we set up a base integration test with validator nodes that can be configured to track a subset of shards. 

We are waiting on the results of in emory trie + single shard tracking experiment to close this ticket.

**Comment by VanBarbascu** [comment]
We have confirmation that nodes tracking a subset of shards are able to download the state from external storage.
Waiting for the monitoring PR to get approved and then we can close this ticket.

**Comment by VanBarbascu** [comment]
The functionality has been finalised. We have the capability to track parts availability using the metrics provided by the dump_check tool. The task of single shard tracking with decentralised state sync will be handled in a separate project.
